src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -1274,42 +1266,49 @@ export default class Blockchain extends Emittery.Typed<
         Buffer /*codeHash*/
       ])[2];
 
-      let keys: Buffer[] = [];
-      return new Promise((resolve, reject) => {
-        storageTrie
-          .createReadStream()
-          .on("data", data => {
-            keys.push(data.key);
-          })
-          .on("end", () => {
-            // #4 - sort and filter keys
-            const sortedKeys = keys.sort((a, b) => Buffer.compare(a, b));
-
-            // find starting point in array of sorted keys
-            const startKeyBuffer = Data.from(startKey).toBuffer();
-            keys = sortedKeys.filter(key => {
-              if (Buffer.compare(startKeyBuffer, key) <= 0) {
-                return key;
-              }
-            });
+      return new Promise<T>((resolve, reject) => {
+        const startKeyBuffer = Data.from(startKey).toBuffer();
+        const compare = (a: Buffer, b: Buffer) => a.compare(b) < 0;
+
+        const keys: Buffer[] = [];
+        const handleData = ({ key }) => {
+          // ignore anything that comes before our starting point
+          if (startKeyBuffer.compare(key) > 0) return;
+
+          // #4 - sort and filter keys
+          // insert the key exactly where it needs to go in the array
+          const position = findInsertPosition(keys, key, compare);
+          // ignore if the value couldn't possibly be relevant
+          if (position > maxResult) return;
+          keys.splice(position, 0, key);
```

Comment:
`findInsertPosition` does a binary search on a _sorted_ array to find the optimal index to insert a new item. We do NOT need to worry about this being a stable sort, since we are sorting the keys, and there will be only one of each key.

There _are_ other data structures better suited for fast insertion, but ultimately we'll need to convert it into an array anyway, so my _guess_ is that an array would end up winning out in practical use anyway.